### ISSUE TITLE
Why doesn't AnnotatedRulesEngine convert beans to a Rule?

### ISSUE THREAD

**Issue by mrcritical** [benas/easy-rules#11]
I am working on capturing business rule events in order to record their execution for debugging purposes. I am wrapping a `Rule` in order to capture the execution. This works well for the `DefaultRulesEngine` but won't work for the `AnnotatedRulesEngine`. It will be tricker to do this with that.

If the `AnnotatedRulesEngine` had converted the annotated object into a `Rule` at registration then this would have been considerably easier to. Does this sound too crazy?

Is there a better way to record the execution (or not) then delegation? We are recording more then just logging. We need to record into a custom reporting library.

**Comment by benas** [comment]
Hi @mrcritical,

Are you trying to provide a custom reporting strategy when a rule is executed? Currently, the engine simply logs a message ("Rule X performed successfully") and you would like to be able to do more than just logging right? I'm not sure I understand your requirement.

Could you please give an example of how it can work with the `DefaultRulesEngine` but not with the `AnnotatedRulesEngine `? Thanks.

Regards
Mahmoud

**Comment by mrcritical** [comment]
Yes, we are essentially doing a custom reporting strategy. Our requirement is to record what rules were executed during an API request. Ultimately we will be recording the results after the business rules as well. This is to help us triage issues that were detected as a result of business rules (not your framework, which is very nice, by the way).

The way I am doing this now is by wrapping a `Rule` and recording an event when `Rule#evaluateConditions() is fired. The event records the rule name, description, and whether it was applied or not. I will be looking into extending this to also record results after `performActions` (but I don't expect your framework to support that, though not sure the best unobtrusive way to do that yet).

I then wrap a `DefaultRulesEngine` to trigger the storage of the recorded rule executions after the`DefaultRulesEngine#fireRules()` call. This makes sure all the executions are stored and available through our reporting interface.

I don't see an easy way to wrap or augment the rules executed by an `AnnotatedRulesEngine` without interacting directly with the rule object. Ideally I would like to avoid that, if at all possible.

**Comment by benas** [comment]
Thanks for these details. I see the point. Easy Rules should provide a new extension point that allows users to provide a custom reporting strategy for each rule. I suggest to implement an interface like:

```java
public interface RuleListener {

    void beforeExecution(Rule rule);

    void afterExecution(Rule rule);

    void onException(Rule rule, Exception exception);
    
}
```

and then register the listener at rule engine construction time. The listener will be called for each registered rule. You would be able to provide a custom reporting strategy in a consistent way between default and annotated rule engines. What do you think about this solution?

**Comment by mrcritical** [comment]
Yes, a listener would be ideal.

**Comment by benas** [comment]
Great! I'll plan that for next release.

Keep tuned.

**Comment by mrcritical** [comment]
Thanks.

**Comment by mrcritical** [comment]
One more question, how will I be able to determine if a rule has been executed or not? Will it be the the triggering of `afterExecution`? Meaning a rule consider will fire `beforeExecution` but will not fire `afterExecution` if it was not actually run. 

I need to be sure to catch if a rule is being considered but may not be executed in addition to rules that are actually executed.

**Comment by benas** [comment]
Yes, if the `afterExecution` method has been triggered, you can be sure that your rule has been executed.

Here is a pseudo code of what the new engine code would be:

```java
if (rule.evaluateConditions()) {
       try {
         ruleListener.beforeExecution(rule);
         rule.performActions();
         ruleListener.afterExecution(rule);
      } catch (Exception exception) {
          ruleListener.onException(rule, exception);
      }
}
```

There is another solution that does not rely on the listener, which is to use a flag in your rule that you set to true when the rule is executed. After firing rules, you can check if the rule has been executed by querying this flag. There is an example of this approach in [here](https://github.com/benas/easy-rules/issues/9).

Hope it helps
Regards

**Comment by mrcritical** [comment]
Either way (or even both) seems good to me. Maybe adding this to the `Rule` interface and, with the proxy, gives a great overall solution. Again, nice improvements. This should make it very easy for us to use your rules engine. It keeps it light weight still, which I love, but just enough power to hook into it. Keep up the great work!

**Comment by benas** [comment]
Thank you Philip!

I'll let you know as soon as I deploy a new snapshot version with all these improvements.

Regards
Mahmoud

**Comment by benas** [comment]
Hi Philip,

I have released version 2.0.0 with the new RuleListener API and [others improvements](http://www.easyrules.org/get-involved/release-notes.html).

Looking forward for your feedback.

Best regards
Mahmoud

**Comment by mrcritical** [comment]
Great, thanks. Will check it out.
